### DISCUSSION TITLE
Training PINN using H^1 Norm

### DISCUSSION THREAD

**Discussion by alanraydan** [lululxvi/deepxde#837] [Q&A] [answered]
Hi all,

I'm using DeepXDE to solve the primitive equations (a collection of 4 PDEs) with two spatial variables and one time variable. Passing the data (PDE residuals and IC/BCs) to `dde.data.TimePDE()` and training the model with this data is equivalent to minimizing the $L^2$ norm between the PINN and the data; however, I am interested in minimizing the $H^1$ norm as well.

For the PDE residuals this isn't a problem since I can just pass the Jacobian of the PDE as another residual, e.g.

`h1_pde_res = dde.grad.jacobian(pde_res, x, i=0, j=0)`

The periodic BCs also aren't hard because I can just use the argument `derivative_order=1`. But for the ICs and Dirichlet BCs there is no option to select higher order derivatives.

Mathematically speaking, given initial data $f(x)$, I would like to train my PINN to minimize the $H^1$ residual $(\partial_x (f(x) - PINN_{\theta}(x, t=0) ))^2$ as well as the $L^2$ residual $(f(x) - PINN_{\theta}(x, t=0) )^2$. Likewise for the Dirichlet BC residuals.

**MAIN QUESTION**: How can I train a PINN to minimize the gradient of the IC residuals and Dirichlet BC residuals rather than just the residuals themselves?

**Comment by forxltk** [comment, marked as answer]
Maybe you can use `Operator BC` instead of `IC` and `Dirichlet BC`. Or define your own `loss function`, and use it for `loss` in `model.compile`

**Reply by alanraydan** [reply]
This worked!

**Reply by inodozi** [reply]
> Maybe you can use `Operator BC` instead of `IC` and `Dirichlet BC`. Or define your own `loss function`, and use it for `loss` in `model.compile`

Hi There
Is it possible to have two different types of metrics for the residuals?
For example, L_2 for the PDE residuals and another metric like Wasserstein distance for the Initial conditions?

**Reply by forxltk** [reply]
Yes, see https://deepxde.readthedocs.io/en/latest/_modules/deepxde/model.html?highlight=loss%20function

**Comment by lululxvi** [comment]
You may be also interested in gPINN https://doi.org/10.1016/j.cma.2022.114823
